sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -111,6 +147,55 @@ private static PipelineResult run(
     return result;
   }
 
+  private static PipelineResult limitRun(
+      PipelineOptions options,
+      BeamRelNode node,
+      DoFn<Row, KV<String, Row>> collectDoFn,
+      DoFn<KV<String, Row>, Void> limitCounterDoFn,
+      LimitStateVar limitStateVar) {
+    ExecutorService pool = Executors.newFixedThreadPool(1);
+
+    options.as(DirectOptions.class).setBlockOnRun(false);
+    Pipeline pipeline = Pipeline.create(options);
+    PCollectionTuple.empty(pipeline)
+        .apply(node.toPTransform())
+        .apply(ParDo.of(collectDoFn))
+        .apply(ParDo.of(limitCounterDoFn));
+
+    PipelineResult result = pipeline.run();
+
+    pool.execute(
```

Comment:
The extra thread seems unnecessary. You can do this in the main thread.

Reply:
fixed.